lib/api/src/modules/url.ts

```diff
@@ -161,6 +178,47 @@ export const init: ModuleFn = ({ store, navigate, state, provider, fullAPI, ...r
       fullAPI.navigateUrl(url, options);
     });
 
+    const initialArgs: Record<string, Story['args']> = {};
+
+    // Track initial args for the currently selected story
+    fullAPI.on(SET_STORIES, () => {
+      const data = fullAPI.getCurrentStoryData();
+      const storyArgs = isStory(data) ? data.args : undefined;
+      initialArgs[data.id] = storyArgs;
+
+      const queryArgs = api.getQueryParam('args') || '';
+      const args = parseArgs(queryArgs);
+      // map args to types of storyArgs
+      fullAPI.updateStoryArgs(data as Story, storyArgs ? { ...storyArgs, ...args } : args);
```

Comment:
Hmm, I think the timing of this is a bit funny. Shouldn't this parsing and setting happen at the same time we set the initial selected story from parsing the URL?